From 04433dedf56da13d6f932f368f7c8e04db113704 Mon Sep 17 00:00:00 2001
From: Antonio Ojea <aojea@google.com>
Date: Thu, 27 Jun 2024 13:27:10 +0200
Subject: [PATCH] netfilter: nfnetlink_queue: unbreak SCTP traffic
Git-commit: 26a77d02891ab62172085a4f94af9b3c90aed387
Patch-mainline: v6.11-rc3
References: bsc#1285684

when packet is enqueued with nfqueue and GSO is enabled, checksum
calculation has to take into account the protocol, as SCTP uses a
32 bits CRC checksum.

Enter skb_gso_segment() path in case of SCTP GSO packets because
skb_zerocopy() does not support for GSO_BY_FRAGS.

Joint work with Pablo.

Signed-off-by: Antonio Ojea <aojea@google.com>
Signed-off-by: Pablo Neira Ayuso <pablo@netfilter.org>

[ alnovak: just adding the nf_queue_checksum_help() helper without the
  sctp branch as a placeholder for further backports ]

Signed-off-by: Ales Novak <alnovak@suse.com>
---
 net/netfilter/nfnetlink_queue.c | 7 ++++++-
 1 file changed, 6 insertions(+), 1 deletion(-)

diff --git a/net/netfilter/nfnetlink_queue.c b/net/netfilter/nfnetlink_queue.c
index e9c50dea69846..992d8e1da01b6 100644
--- a/net/netfilter/nfnetlink_queue.c
+++ b/net/netfilter/nfnetlink_queue.c
@@ -221,6 +221,11 @@ nfqnl_flush(struct nfqnl_instance *queue, nfqnl_cmpfn cmpfn, unsigned long data)
 	spin_unlock_bh(&queue->lock);
 }
 
+static int nf_queue_checksum_help(struct sk_buff *entskb)
+{
+	return skb_checksum_help(entskb);
+}
+
 static struct sk_buff *
 nfqnl_build_packet_message(struct nfqnl_instance *queue,
 			   struct nf_queue_entry *entry)
@@ -257,7 +262,7 @@ nfqnl_build_packet_message(struct nfqnl_instance *queue,
 
 	case NFQNL_COPY_PACKET:
 		if (entskb->ip_summed == CHECKSUM_PARTIAL &&
-		    skb_checksum_help(entskb))
+		    nf_queue_checksum_help(entskb))
 			return NULL;
 
 		data_len = ACCESS_ONCE(queue->copy_range);
-- 
2.56.0

